core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -1698,6 +1699,113 @@ RexNode simplifyAnd2ForUnknownAsFalse(List<RexNode> terms,
     }
   }
 
+  /**
+   * <p>Simplifies AND/OR condition that has common expressions,
+   * extracts and eliminates/merges them as much as possible.</p>
+   * <ul>
+   * <li>(a OR b) AND (a OR b OR c OR d) =&gt; (a OR b)
+   * <li>(a OR b OR c OR d) AND (a OR b OR e OR f) =&gt; (a OR b) OR ((c OR d) AND (e OR f))
+
+   * <li>(a AND b) OR (a AND b AND c) =&gt; (a AND b)
+   * <li>(a AND b AND c AND d) OR (a AND b AND e AND f) =&gt;
+   * (a AND b) AND ((c AND d) OR (e AND f))
+   * </ul>
+   * <p> The difference between {@link #simplifyAnd} is that {@link #simplifyAnd} mainly
+   * simplifies expressions whose answer can be determined without evaluating both sides,
+   * like: FALSE AND (xxx) =&gt; FALSE</p>
+   */
+  public RexNode eliminateCommonExprInCondition(RexNode node) {
+    if (!(node instanceof RexCall)) {
+      return node;
+    }
+    // Simplify children recursively
+    // so that when we simplify parent node
+    // the child node should be simplified
+    List<RexNode> operands = ((RexCall) node).getOperands();
+    List<RexNode> simplifiedChildren = operands.stream()
+        .map(this::eliminateCommonExprInCondition)
+        .collect(Collectors.toList());
+
+    switch (node.getKind()) {
+    case AND: {
+      // Split the children to get the disjunctive predicates.
+      List<List<RexNode>> disjunctions = simplifiedChildren.stream()
+          .map(RelOptUtil::disjunctions)
+          .collect(Collectors.toList());
+
+      // Find the common predict among the list by retainAll.
+      List<RexNode> common = new ArrayList<>(disjunctions.get(0));
+      disjunctions.forEach(common::retainAll);
+
+      if (common.isEmpty()) {
+        // no common, return the original but with child simplified
+        return and(simplifiedChildren);
+      }
+
+      // Get the predicates that without the common part.
+      disjunctions.forEach(p -> p.removeAll(common));
+
+      if (disjunctions.stream().anyMatch(List::isEmpty)) {
+        // (a || b) && (a || b || c || d) => (a || b)
+        return or(common);
+      }
+      // (a || b || c || d) && (a || b || e || f) =>
+      // (a || b) || ((c || d) && (e || f))
+      List<RexNode> splice = disjunctions.stream()
+          .map(this::or)
+          .collect(Collectors.toList());
+      return or(or(common), and(splice));
```

Comment:
`or(or(...), ...)` seems to be a sub-optimal representation. Could you flatten it?

Reply:
`or(or(...), ...)`'s advance is more readable, I found `RexUtil.composeDisjunction` will o the flatten anyway(flattenOr), do you think we can keep this here?